### PR TITLE
:seedling: pkg/../types_apiexport: add name, namespace

### PR DESCRIPTION
**Pull request by s-urbaniak**
## Summary
This adds name,namespace fields to permission claims so we can work independently on reverse permission claim authz and narrowly scoped permissionclaims by @nrb.

## Related issue(s)

Relates to #2089
Relates to #1937

### CODE REVIEW

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 1]
```diff
@@ -188,6 +188,17 @@ const (
 type PermissionClaim struct {
 	GroupResource `json:","`
 
+	// name of an object within the group/resource.
+	// It matches the metadata.name field of the underlying object.
+	// If unset, all objects having the group/resource are being claimed.
+	// +optional
+	Name string `json:"name,omitempty"`
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:195]: Indeed, we didn't 🤔 Are double entries allowed with the same group/resources/identityhash?

[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:195]: @nrb i guess this case is missing in https://docs.google.com/document/d/1IHai7FlVQXPCD_MZObT0CzAkdOFSZuWGokHa_kn72HQ/edit as well.

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 2]
```diff
@@ -196,6 +201,19 @@ type PermissionClaim struct {
 	IdentityHash string `json:"identityHash,omitempty"`
 }
 
+type ClaimedResource struct {
+	// name of an object within a claimed group/resource.
+	// It matches the metadata.name field of the underlying object.
+	// If unset, all objects having the group/resource are being claimed.
+	// +optional
+	Name string `json:"name,omitempty"`
```
Comment (stevekuznetsov) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:209]: Can we (trivially) add validation for these fields to match what an object's name would have to be?

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 3]
```diff
@@ -196,6 +201,19 @@ type PermissionClaim struct {
...
+
+	// namespace containing the named object. Matches metadata.namespace field.
+	// If namespace is set but name is not set this field is ignored.
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:212]: this is strange. How can I claim a complete namespace?

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 4]
```diff
@@ -188,6 +188,11 @@ const (
 type PermissionClaim struct {
 	GroupResource `json:","`
 
+	// claimed is a list of claimed resources.
+	// If omitted all resources of the given group/resource will be claimed.
+	// +optional
+	Claimed []ClaimedResource `json:"claimed,omitempty"`
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:194]: resourceSelector

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:194]: later:
```
resourceSelector:
  name:
  namespace:
  labels:
  jsonPath:
```

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:194]: What is default? What do we want to encourage?

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:194]:
```
resourceSelector:
- name:
  namespace:
  labels:
  jsonPath:
  everything: true // this is requires all others are unset, and only length 1 is allowed for resourceSelector
```

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:194]: + required for resourceSelector

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:194]: alternative:
```
resourceSelector: []
everything: true
```

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:209]: sorry i am not following 🙈 can you elaborate? do we have restrictions on names?

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nrb) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:194]: I like `resourceSelector`. I also think that `name/namespace`, `labels`, and `jsonPath` should all be mutually exclusive.

Given the security implications, requiring that `everything` is explicitly defined makes sense to me, too.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 3 above)
Comment (nrb) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:212]: As I understand it, it's for scoping named resources to a namespace.

I suppose one could try to claim all instances of the RG in a namespace, but so far I haven't thought of or seen a use case for that.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nrb) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:209]: You're right, it is missing.

I could see it being possible in the case of ConfigMaps - a developer may want to only see 2 or 3 ConfigMaps, or perhaps 2-3 secrets.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 2 above)
Comment (nrb) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:209]: I think we'd have to ensure we match one or more of these restrictions: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names

tl;dr - if we want to put the names in labels or paths, we'll need to make sure that the name provided follows those constraints.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go]: @sttts it was a suggestion from your side when we did a quick chat. We can remove it if it doesn't make sense though.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go]: @nrb I called the "everything" field `all` for the sake of shortness, wdyt?

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:217]: hrm, i believe this would indeed need to be checked during admission. let me see if this can be done declaratively.

Comment (stevekuznetsov) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:217]: If we can't trivially do it via some kube-builder annotation or something for validation, maybe not worth it.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nrb) [reply pkg/apis/apis/v1alpha1/types_apiexport.go]: Makes sense to me

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 2 above)
Comment (nrb) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:217]: The closest I can see with the kube-builder annotations (defined [here](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#schemaObject) would be a combo on maxLength and a regex pattern.

[API Machinery has some validation helpers](https://github.com/kubernetes/kubernetes/blob/c18c2447cf602ee1a76113200b3f5b6200a24a34/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go) that appear to be used within Kube itself.

**Comment by s-urbaniak** [comment]
> CI / e2e-sharded - Failing

I cannot reproduce this locally 🤔

**Comment by stevekuznetsov** [comment]
Does it fail 100% in CI?

**Comment by s-urbaniak** [comment]
@stevekuznetsov so far twice but only on github 🤔 A local `ARTIFACT_DIR=$PWD/artifacts UNSAFE_E2E_HACK_DISABLE_ETCD_FSYNC=true E2E_PARALLELISM=8 make test-e2e-sharded WHAT=./test/e2e/virtual/apiexport TEST_ARGS="-run TestAPIExportAPIBindingsAccess -v"` yielded a success.
```

```

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 5]
```diff
@@ -197,6 +210,20 @@ type PermissionClaim struct {
...
+	// name of an object within a claimed group/resource.
+	// It matches the metadata.name field of the underlying object.
+	//
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// namespace containing the named object. Matches metadata.namespace field.
+	// If "name" is unset, all objects from the namespace are being claimed.
+	//
+	// +required
```
Comment (nrb) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:223]: Is this required if we use a label selector? Or are you intending for `resourceSelector` and label selection to be mutually exclusive?

Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:223]: This won't be required if we use a label selector. For now, the constrained declared here is ok but once we add label selection into this struct, I guess we'll need CEL to express mutual exclusivity.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:218]: updated with the referenced regex but fyi: I just found a second ago a precedence for this in our code 😛 https://github.com/kcp-dev/kcp/blob/main/pkg/apis/apis/v1alpha1/types_apiexport.go#L229

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 5 above)
Comment (nrb) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:225]: Kube has (or had, not up-to-date on the migration) a solution for this on the `VolumeSource` of a PV. [See this comment](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L48).

I'll dig around some more if I can see the exact mechanism, I don't think we need CEL for this unless it's the new way of doing things.

**Comment by stevekuznetsov** [comment]
CI is unhappy, #2278 , below

```
 1 files have incorrect boilerplate headers:
pkg/apis/test/pattern.go 
```

```
 go module files are out of date 
```

**Comment by stevekuznetsov** [comment]
Flake on #2281 
/retest
/lgtm
/approve

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 5 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:225]: I have precedence of mutual exlusitivity using CEL here: https://github.com/kcp-dev/kcp/pull/2222/files#diff-228a80bf14ca2df5c0d6053a9a4ed5f5ddb5d74a13c2618584741e2723afbacbR190
```
// +kubebuilder:validation:XValidation:rule="self.all != has(self.resourceSelector)",message="either \"all\" or \"resourceSelector\" must be set"
type PermissionClaim struct {
```

Comment (nrb) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:225]: Awesome. I didn't see the `XValidation` marker when looking at kubebuilder markers, but that's perfect.

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 6]
```diff
@@ -186,9 +186,22 @@ const (
...
+//
+// +kubebuilder:validation:XValidation:rule="self.all != has(self.resourceSelector)",message="either \"all\" or \"resourceSelector\" must be set"
 type PermissionClaim struct {
 	GroupResource `json:","`
 
+	// all claims all resources for the given group/resource.
+	// This is mutually exclusive with resourceSelector.
+	// +required
+	// +kubebuilder:default=false
+	All bool `json:"all"`
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:198]: why not omitempty? IMO having all all the time is ugly.

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 7]
```diff
@@ -186,9 +186,22 @@ const (
 // PermissionClaim identifies an object by GR and identity hash.
 // Its purpose is to determine the added permissions that a service provider may
 // request and that a consumer may accept and allow the service provider access to.
+//
+// +kubebuilder:validation:XValidation:rule="self.all != has(self.resourceSelector)",message="either \"all\" or \"resourceSelector\" must be set"
 type PermissionClaim struct {
 	GroupResource `json:","`
 
+	// all claims all resources for the given group/resource.
+	// This is mutually exclusive with resourceSelector.
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:195]: where is this checked?

File: pkg/apis/apis/v1alpha1/types_apiexport.go [hunk 8]
```diff
@@ -197,6 +210,22 @@ type PermissionClaim struct {
...
+	// +optional
+	// +kubebuilder:validation:Pattern="^([a-z0-9][-a-z0-9_.]*)?[a-z0-9]$"
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name,omitempty"`
+
+	// namespace containing the named object. Matches metadata.namespace field.
+	// If "name" is unset, all objects from the namespace are being claimed.
+	//
+	// +required
+	Namespace string `json:"namespace"`
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apiexport.go:226]: why not omit empty?

Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:226]: and why not optional?

Comment (sttts) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:226]: being required does not extend to more selector types.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 6 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: fixed, set `+optional` and `omitemtpy`

Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:198]: note: `omitempty` implies `+optional`. I debugged half of a day in controller-tools to find out.

File: pkg/apis/apis/v1alpha1/types_apiexport.go (hunk 8 above)
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apiexport.go:226]: fixed, however how do we want to validate this, using CEL?
